File: projects/swagger-studio-ruleset/typescript/src/_http.ts
/**
 * Shared HTTP helpers — every Studio-touching module uses the same headers,
 * base URL, and timeout machinery. Centralizing keeps drift out of the
 * activator/deleter/lister/puller surfaces.
 */

import type { Settings } from "./config.js";

export function studioBaseUrl(settings: Settings): string {
  return settings.swaggerhubBaseUrl.replace(/\/$/, "");
}

export function buildHeaders(
  settings: Settings,
  overrides: Record<string, string> = {},
): Record<string, string> {
  return {
    Authorization: settings.swaggerhubApiKey,
    Accept: "application/json",
    "User-Agent": "api-foundation-ruleset-publisher/0.1",
    ...overrides,
  };
}

/**
 * Run `fn` against a fetch that aborts after the configured timeout.
 * Always clears the timer so it doesn't keep the event loop alive.
 */
export async function withTimeout<T>(
  settings: Settings,
  fn: (signal: AbortSignal) => Promise<T>,
): Promise<T> {
  const controller = new AbortController();
  const timer = setTimeout(() => {
    controller.abort();
  }, settings.publisherRequestTimeoutMs);
  try {
    return await fn(controller.signal);
  } finally {
    clearTimeout(timer);
  }
}

File: projects/swagger-studio-ruleset/python/src/swagger_studio_ruleset_publisher/_http.py
"""Shared HTTPS client factory.

All Studio-touching modules (activator, deleter, lister, puller) use the
same auth, base URL, timeout, and user agent. Centralizing the factory
keeps drift from creeping in across modules.
"""

from __future__ import annotations

import httpx

from swagger_studio_ruleset_publisher.config import Settings


def create_client(settings: Settings, accept: str = "application/json") -> httpx.AsyncClient:
    """Build an httpx.AsyncClient configured for the SwaggerHub Standardization API.

    `accept` is overridable so binary fetchers (puller) can request `application/zip`.
    """
    return httpx.AsyncClient(
        base_url=settings.swaggerhub_base_url,
        timeout=settings.publisher_request_timeout_s,
        headers={
            "Authorization": settings.swaggerhub_api_key.get_secret_value(),
            "Accept": accept,
            "User-Agent": "api-foundation-ruleset-publisher/0.1",
        },
    )
